broker-daemon/utils/migrate-store.js

```diff
@@ -1,47 +1,65 @@
+const { promisify } = require('util')
+
+// `previousBatch` allows us to batch save records in sequence and lock the operation
+// to make sure we are not inserting records out of the order they are received.
+//
+// We set this value to automatically resolve on the first use. This value is then
+// subsequently set when a batch needs to be processed
+let previousBatch
+
 /**
  * Pipe data from one sublevel store into another
+ *
  * @todo Refactor into an iterator pattern using streams2
  * @param  {sublevel} sourceStore       - Sublevel that is the source of the data to pipe
  * @param  {sublevel} targetStore       - Sublevel that is the destination of the data
  * @param  {Function} createDbOperation - Function that returns a sublevel batch-compatible database operation to take on the target store
  * @param  {number}   batchSize         - Number of operations to batch before executing on the target database
  * @returns {Promise<void>}
  */
-async function migrateStore (sourceStore, targetStore, createDbOperation, batchSize = 20) {
+async function migrateStore (sourceStore, targetStore, createDbOperation, batchSize = 2000) {
   return new Promise((resolve, reject) => {
     const stream = sourceStore.createReadStream()
+
     let batch = []
 
-    function flush (done = () => {}) {
-      if (!batch.length) {
-        return process.nextTick(done)
+    stream.on('error', reject)
+
+    stream.on('end', async () => {
+      try {
+        if (previousBatch) {
+          await previousBatch()
+        }
+        return resolve()
+      } catch (e) {
+        return reject(e)
       }
+    })
 
-      targetStore.batch(batch, (err) => {
-        if (err) return reject(err)
-        done()
-      })
+    stream.on('data', async ({ key, value }) => {
+      // The first time we start we should initialize the previousBatch variable
+      if (!previousBatch) {
+        previousBatch = async () => {}
```

Comment:
I don't think you need to do this.

I think:

```
let myPromise
await myPromise
```

will resolve on the next tick.

Reply:
@treygriffith that is not the functionality that im seeing. You have to call the function don't you (I added this in an earlier comment). I've created a small example, does this look correct?

```
async function dan () {
  console.log('ok')
}

async function sam () {
  await dan
}

sam()
```

There will be no log that is generated

Reply:
that's because `dan` in this example is a function that returns a promise, not a promise.

`previousBatch` by contrast is the promise itself.

```
const writeBatch = promisify(store.batch)
previousBatch = writeBatch(batch)

...
await previousBatch
```

Reply:
@treygriffith that was it. ty for that